File: lurien/src/snapshot.js
// The page as a list of addressable nodes, evaluated in the page.
//
// Loaded after `locator.js`, whose role table, name computation and visibility
// rules this reuses: a snapshot that named elements differently from the way a
// selector finds them would be a second convention, and a handle taken from one
// would not resolve through the other.
//
// A node carries a `path` (a CSS path) and a handle `eN`. The driver keeps the
// handle table, so the page is never tagged with an attribute a script could
// read, and a handle is checked against the role and name it was captured with
// before it is acted on.
//
// Returns a JSON string.

// Elements worth an entry even when they hold no text: they are what an act
// verb can address.
const LURIEN_ADDRESSABLE =
  "a[href], area[href], button, input, select, textarea, summary, option, " +
  "progress, output, dialog, [role], [contenteditable=true], [contenteditable='']";

// Text long enough to be prose is cut: a snapshot is a map of the page, and an
// agent that wants the article reads it with `text`.
const LURIEN_TEXT_CAP = 160;

// Regions worth a line even when they carry no text of their own: they are what
// makes the list a map of the page rather than a pile of controls.
const LURIEN_STRUCTURE = new Set([
  "form",
  "search",
  "navigation",
  "main",
  "banner",
  "contentinfo",
  "dialog",
  "table",
  "list",
  "tabpanel",
]);

// The role a snapshot line carries. An element with no role of its own that is
// listed for its text is `text`, and both the walk and the handle check read
// this, or a text node's handle would never verify.
function lurienEntryRole(el) {
  const role = lurienNodeRole(el);
  const known = LURIEN_ROLES[role] !== undefined && role !== "paragraph";
  return known || el.matches(LURIEN_ADDRESSABLE) ? role : "text";
}

function lurienNodeState(el, role) {
  const state = [];
  if (el.disabled || el.getAttribute("aria-disabled") === "true") {
    state.push("disabled");
  }
  if (role === "checkbox" || role === "radio" || role === "switch") {
    const on = el.checked === undefined ? el.getAttribute("aria-checked") === "true" : el.checked;
    state.push(on ? "checked" : "unchecked");
  }
  if (el.getAttribute("aria-expanded")) {
    state.push("expanded=" + el.getAttribute("aria-expanded"));
  }
  if (role === "heading") {
    const level = el.getAttribute("aria-level") || el.localName.slice(1);
    if (/^[1-9]$/.test(level)) {
      state.push("level=" + level);
    }
  }
  if (el.localName === "input" || el.localName === "textarea") {
    const type = (el.getAttribute("type") || "text").toLowerCase();
    if (type === "file") {
      state.push("file");
    }
    if (el.required) {
      state.push("required");
    }
  }
  return state;
}

// The text a node contributes on its own, which is only the text no child of it
// already reports.
function lurienOwnText(el) {
  let own = "";
  for (const node of el.childNodes) {
    if (node.nodeType === 3) {
      own += node.nodeValue;
    }
  }
  return LURIEN_NORM(own);
}

/**
 * The page as an ordered node list.
 *
 * `limit` bounds the list, because a snapshot an agent pays for by the token
 * must have a ceiling: the answer says how many nodes were dropped rather than
 * pretending the page ended.
 */
function lurienSnapshot(limit) {
  const nodes = [];
  let dropped = 0;
  let counter = 0;

  const walk = (el, depth, carried) => {
    let mine = depth;
    let held = carried;
    const role = lurienNodeRole(el);
    const addressable = el.matches(LURIEN_ADDRESSABLE);
    const name = LURIEN_NORM(lurienName(el)).slice(0, LURIEN_TEXT_CAP);
    const own = lurienOwnText(el).slice(0, LURIEN_TEXT_CAP);
    const known = LURIEN_ROLES[role] !== undefined && role !== "paragraph";
    // A label whose text is already the name of its control says nothing twice.
    const label = el.localName === "label" && el.control &&
      LURIEN_LOWER(lurienName(el.control)) === LURIEN_LOWER(name || own);
    // A span inside a button is not a second node: the button already reports
    // that text, and an agent charged per token should not read it twice.
    const echo = !addressable && !known && own && carried.includes(LURIEN_LOWER(own));
    const region = LURIEN_STRUCTURE.has(role);
    const worth = !label && (addressable || region || (known && (name || own)) || (own && !echo));
    if (worth && lurienVisible(el)) {
      if (nodes.length >= limit) {
        dropped += 1;
      } else {
        counter += 1;
        const entry = {
          ref: "e" + counter,
          role: lurienEntryRole(el),
          name: name || own,
          path: lurienPath(el),
          depth,
          state: lurienNodeState(el, role),
        };
        const value = el.value;
        if (typeof value === "string" && value && el.type !== "password") {
          entry.value = value.slice(0, LURIEN_TEXT_CAP);
        }
        nodes.push(entry);
        mine = depth + 1;
        held = LURIEN_LOWER(entry.name);
      }
    }
    for (const child of el.children) {
      walk(child, mine, held);
    }
  };

  if (document.body) {
    walk(document.body, 0, "");
  }
  return JSON.stringify({
    title: document.title,
    url: location.href,
    nodes,
    dropped,
  });
}

/**
 * Is the handle's element still the element the handle was captured for?
 *
 * A handle is a promise about one node, and a page that re-rendered underneath
 * it can leave the path pointing at something else. Checking role and name
 * turns that into a refusal instead of a click on whatever moved into place.
 */
function lurienVerify(path, role, name) {
  let el = null;
  try {
    el = document.querySelector(path);
  } catch (e) {
    return JSON.stringify({ ok: false, why: "invalid", detail: String(e && e.message) });
  }
  if (!el) {
    return JSON.stringify({ ok: false, why: "gone" });
  }
  const now = lurienEntryRole(el);
  const label = LURIEN_NORM(lurienName(el)) || lurienOwnText(el);
  if (now !== role || LURIEN_LOWER(label) !== LURIEN_LOWER(name)) {
    return JSON.stringify({ ok: false, why: "changed", role: now, name: label });
  }
  return JSON.stringify({ ok: true });
}

File: lurien/src/snapshot.rs
//! The page as a list of addressable nodes, and the handles that address them.
//!
//! An agent pays for a page by the token. Page source is the worst possible
//! representation of it: mostly markup it cannot act on, and different after
//! every redesign even when the page still does the same thing. So the default
//! snapshot is a role/name list, in document order, with one handle per node:
//!
//! ```text
//! - heading "Sign in" [level=1] [ref=e1]
//! - textbox "Email" [ref=e2]
//! - button "Log in" [ref=e3]
//! ```
//!
//! A handle goes back to any verb that takes a selector, as `ref:e3`. The
//! handle table lives in the driver, so nothing in the page is tagged, and a
//! handle is checked against the role and name it was captured with before it is
//! acted on: a page that re-rendered underneath a handle earns a refusal, not a
//! click on whatever moved into place.

use crate::error::Error;
use runtime_foxdriver::Page;

/// The walker, evaluated in the page after the resolver it borrows from.
const WALKER: &str = include_str!("snapshot.js");

/// How many nodes a snapshot reports before it says it stopped.
pub const DEFAULT_LIMIT: usize = 200;

/// One addressable node.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Node {
    /// The handle a caller sends back, without its `ref:` prefix.
    pub handle: String,
    /// Role, explicit or implicit.
    pub role: String,
    /// Accessible name, or the node's own text when it has no name.
    pub name: String,
    /// CSS path, used to act and to check the handle still means this node.
    pub path: String,
    /// Nesting depth among reported nodes.
    pub depth: usize,
    /// Flags worth a token: `disabled`, `checked`, `level=2`, `required`.
    pub state: Vec<String>,
    /// Current value of a control, when it has one and is not a password.
    pub value: Option<String>,
}

/// One capture of a page.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Snapshot {
    /// Document title.
    pub title: String,
    /// Document URL.
    pub url: String,
    /// Nodes in document order.
    pub nodes: Vec<Node>,
    /// Nodes past the limit, reported rather than hidden.
    pub dropped: usize,
}

/// Walk the page and assign a handle to every addressable node.
pub async fn capture(page: &Page, limit: usize) -> Result<Snapshot, Error> {
    let script = format!(
        "(() => {{ {resolver}\n{walker}\n return lurienSnapshot({limit}); }})()",
        resolver = crate::locator::RESOLVER,
        walker = WALKER,
    );
    let answer: String = page
        .evaluate(&script)
        .await
        .map_err(|e| Error::Other(e.to_string()))?
        .into_value()
        .map_err(|e| Error::Other(format!("snapshot did not answer: {e}")))?;
    parse(&answer)
}

/// Does `node`'s handle still point at the node it was captured for?
///
/// The failure is deliberately not "no such element": a stale handle and a
/// missing element need different corrections, and only one of them is fixed by
/// taking a new snapshot.
pub async fn verify(page: &Page, node: &Node) -> Result<(), Error> {
    let script = format!(
        "(() => {{ {resolver}\n{walker}\n return lurienVerify({path}, {role}, {name}); }})()",
        resolver = crate::locator::RESOLVER,
        walker = WALKER,
        path = json_string(&node.path),
        role = json_string(&node.role),
        name = json_string(&node.name),
    );
    let answer: String = page
        .evaluate(&script)
        .await
        .map_err(|e| Error::Other(e.to_string()))?
        .into_value()
        .map_err(|e| Error::Other(format!("handle check did not answer: {e}")))?;
    let answer: serde_json::Value =
        serde_json::from_str(&answer).map_err(|e| Error::Other(e.to_string()))?;
    if answer["ok"].as_bool() == Some(true) {
        return Ok(());
    }
    let detail = match answer["why"].as_str().unwrap_or("gone") {
        "changed" => format!(
            "the page changed under the handle: {} {:?} is now {} {:?}",
            node.role,
            node.name,
            answer["role"].as_str().unwrap_or("?"),
            answer["name"].as_str().unwrap_or("")
        ),
        "invalid" => format!(
            "the handle's path no longer parses ({})",
            answer["detail"].as_str().unwrap_or("rejected by the page")
        ),
        _ => format!("the {} {:?} it named is gone", node.role, node.name),
    };
    Err(Error::Unresolved {
        selector: format!("ref:{}", node.handle),
        detail,
        waited_ms: 0,
        action: "take a fresh snapshot and use the handle it reports".to_string(),
    })
}

impl Snapshot {
    /// The representation an agent reads: one line per node, indented by depth.
    #[must_use]
    pub fn render(&self) -> String {
        let mut out = format!("title: {}\nurl: {}", self.title, self.url);
        for node in &self.nodes {
            out.push('\n');
            for _ in 0..node.depth {
                out.push_str("  ");
            }
            out.push_str("- ");
            out.push_str(&node.role);
            if !node.name.is_empty() {
                out.push_str(&format!(" {:?}", node.name));
            }
            if let Some(value) = &node.value {
                out.push_str(&format!(" [value={value:?}]"));
            }
            for flag in &node.state {
                out.push_str(&format!(" [{flag}]"));
            }
            out.push_str(&format!(" [ref={}]", node.handle));
        }
        if self.dropped > 0 {
            out.push_str(&format!(
                "\n\n{} more node(s) past the limit. Raise `limit` to see them.",
                self.dropped
            ));
        }
        out.push('\n');
        out
    }

    /// The node a handle names, with or without its `ref:` prefix.
    #[must_use]
    pub fn node(&self, handle: &str) -> Option<&Node> {
        let handle = handle.strip_prefix("ref:").unwrap_or(handle);
        self.nodes.iter().find(|n| n.handle == handle)
    }

    /// Handles a caller can use, for the error when one is unknown.
    #[must_use]
    pub fn handles(&self) -> String {
        match (self.nodes.first(), self.nodes.last()) {
            (Some(first), Some(last)) if first.handle != last.handle => {
                format!("{}..{}", first.handle, last.handle)
            }
            (Some(only), _) => only.handle.clone(),
            _ => "none: the snapshot was empty".to_string(),
        }
    }
}

fn parse(raw: &str) -> Result<Snapshot, Error> {
    let value: serde_json::Value =
        serde_json::from_str(raw).map_err(|e| Error::Other(format!("snapshot is not JSON: {e}")))?;
    let nodes = value["nodes"]
        .as_array()
        .map(|rows| rows.iter().map(node_from).collect())
        .unwrap_or_default();
    Ok(Snapshot {
        title: value["title"].as_str().unwrap_or_default().to_string(),
        url: value["url"].as_str().unwrap_or_default().to_string(),
        nodes,
        dropped: value["dropped"].as_u64().unwrap_or(0) as usize,
    })
}

fn node_from(row: &serde_json::Value) -> Node {
    Node {
        handle: row["ref"].as_str().unwrap_or_default().to_string(),
        role: row["role"].as_str().unwrap_or("generic").to_string(),
        name: row["name"].as_str().unwrap_or_default().to_string(),
        path: row["path"].as_str().unwrap_or_default().to_string(),
        depth: row["depth"].as_u64().unwrap_or(0) as usize,
        state: row["state"]
            .as_array()
            .map(|flags| {
                flags
                    .iter()
                    .filter_map(|f| f.as_str().map(str::to_string))
                    .collect()
            })
            .unwrap_or_default(),
        value: row["value"].as_str().map(str::to_string),
    }
}

fn json_string(raw: &str) -> String {
    serde_json::Value::String(raw.to_string()).to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    fn page() -> Snapshot {
        parse(
            r#"{"title":"Sign in","url":"https://example.com/login","dropped":3,"nodes":[
              {"ref":"e1","role":"heading","name":"Sign in","path":"h1","depth":0,"state":["level=1"]},
              {"ref":"e2","role":"textbox","name":"Email","path":"form > input","depth":1,
               "state":["required"],"value":"me@example.com"},
              {"ref":"e3","role":"button","name":"Log in","path":"form > button","depth":1,
               "state":["disabled"]}
            ]}"#,
        )
        .expect("the fixture is JSON")
    }

    /// A snapshot is the representation an agent acts from, so every line has to
    /// carry the three things an act needs: what the node is, what it is called,
    /// and how to address it.
    #[test]
    fn every_line_names_a_role_a_name_and_a_handle() {
        let rendered = page().render();
        for line in rendered.lines().filter(|l| l.trim_start().starts_with("- ")) {
            assert!(
                line.contains("[ref=e"),
                "a node with no handle cannot be acted on: {line}"
            );
        }
        assert!(rendered.contains("- heading \"Sign in\" [level=1] [ref=e1]"), "{rendered}");
        assert!(
            rendered.contains("  - textbox \"Email\" [value=\"me@example.com\"] [required] [ref=e2]"),
            "{rendered}"
        );
        assert!(rendered.contains("  - button \"Log in\" [disabled] [ref=e3]"), "{rendered}");
    }

    /// A page cut off at the limit must say so. A snapshot that silently ends is
    /// read as "the page has nothing else", which is how an agent concludes a
    /// button does not exist.
    #[test]
    fn a_truncated_snapshot_says_what_it_dropped() {
        let rendered = page().render();
        assert!(rendered.contains("3 more node(s) past the limit"), "{rendered}");
        let whole = Snapshot { dropped: 0, ..page() };
        assert!(!whole.render().contains("past the limit"));
    }

    /// The handle is the whole point of the representation: it has to come back
    /// in either form a caller might send it.
    #[test]
    fn a_handle_resolves_with_or_without_its_prefix() {
        let snap = page();
        assert_eq!(snap.node("e2").map(|n| n.role.as_str()), Some("textbox"));
        assert_eq!(snap.node("ref:e2").map(|n| n.path.as_str()), Some("form > input"));
        assert!(snap.node("e99").is_none());
        assert_eq!(snap.handles(), "e1..e3");
    }

    /// Depth is what makes the list a map of the page rather than a pile of
    /// controls: a caller reads which nodes belong to which region.
    #[test]
    fn nesting_is_indented() {
        let rendered = page().render();
        let email = rendered
            .lines()
            .find(|l| l.contains("\"Email\""))
            .expect("the field is rendered");
        assert!(email.starts_with("  - "), "a nested node is indented: {email:?}");
        let heading = rendered
            .lines()
            .find(|l| l.contains("\"Sign in\""))
            .expect("the heading is rendered");
        assert!(heading.starts_with("- "), "a top-level node is not: {heading:?}");
    }

    /// An empty page is a legitimate answer and must not read as an error or
    /// invent a handle.
    #[test]
    fn an_empty_page_renders_its_title_and_nothing_else() {
        let snap = parse(r#"{"title":"","url":"about:blank","nodes":[],"dropped":0}"#)
            .expect("the fixture is JSON");
        assert_eq!(snap.render(), "title: \nurl: about:blank\n");
        assert_eq!(snap.handles(), "none: the snapshot was empty");
    }
}
